File: module-app/src/main/java/maple/expectation/dto/v5/EquipmentExpectationResponseV5.java
package maple.expectation.dto.v5;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;
import lombok.Builder;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

/**
 * V5 CQRS 장비 기대값 응답 DTO
 *
 * @deprecated Kotlin으로 마이그레이션되었습니다. {@code
 *     maple.expectation.web.dto.v5.EquipmentExpectationResponseV5}를 사용하세요.
 */
@Deprecated(since = "2026-02-28", forRemoval = true)
@Getter
@Builder
@Jacksonized
public class EquipmentExpectationResponseV5 {

  // ==================== 기본 정보 ====================

  private final String userIgn;
  private final Instant calculatedAt;
  private final boolean fromCache;

  // ==================== 최대 기대값 프리셋 ====================

  private final BigDecimal totalExpectedCost; // 3개 프리셋 중 최대 기대값
  private final String totalCostText; // "12조 3456억 7890만"
  private final CostBreakdownDto totalCostBreakdown;
  private final int maxPresetNo; // 최대 기대값 프리셋 번호 (1, 2, 3)

  // ==================== 프리셋별 기대값 ====================

  private final List<PresetExpectation> presets;

  // ==================== 내부 DTO ====================

  @Getter
  @Builder
  @Jacksonized
  public static class PresetExpectation {
    private final int presetNo;
    private final BigDecimal totalExpectedCost;
    private final String totalCostText;
    private final CostBreakdownDto costBreakdown;
    private final List<ItemExpectationV5> items;
  }

  @Getter
  @Builder
  @Jacksonized
  public static class ItemExpectationV5 {
    private final String itemName;
    private final String itemIcon;
    private final String itemPart;
    private final int itemLevel;
    private final BigDecimal expectedCost;
    private final String expectedCostText;
    private final CostBreakdownDto costBreakdown;
    private final String enhancePath;

    // 잠재능력 정보
    private final String potentialGrade;
    private final String additionalPotentialGrade;

    // 스타포스 정보
    private final int currentStar;
    private final int targetStar;
    private final boolean isNoljang;

    // 특수 스킬 반지 레벨
    private final int specialRingLevel;

    // 큐브별 기대값
    private final CubeExpectationDto blackCubeExpectation;
    private final CubeExpectationDto additionalCubeExpectation;

    // 스타포스 옵션별 기대값
    private final StarforceExpectationDto starforceExpectation;

    // 환생의 불꽃 기대값
    private final FlameExpectationDto flameExpectation;
  }

  @Getter
  @Builder
  @Jacksonized
  public static class CubeExpectationDto {
    private final BigDecimal expectedCost;
    private final String expectedCostText;
    private final BigDecimal expectedTrials;
    private final String currentGrade;
    private final String targetGrade;
    private final String potential;

    public static CubeExpectationDto empty() {
      return CubeExpectationDto.builder()
          .expectedCost(BigDecimal.ZERO)
          .expectedCostText("0")
          .expectedTrials(BigDecimal.ZERO)
          .potential("")
          .build();
    }
  }

  @Getter
  @Builder
  @Jacksonized
  public static class StarforceExpectationDto {
    private final int currentStar;
    private final int targetStar;
    private final boolean isNoljang;
    private final BigDecimal costWithoutDestroyPrevention;
    private final String costWithoutDestroyPreventionText;
    private final BigDecimal expectedDestroyCountWithout;
    private final BigDecimal costWithDestroyPrevention;
    private final String costWithDestroyPreventionText;
    private final BigDecimal expectedDestroyCountWith;

    public static StarforceExpectationDto empty() {
      return StarforceExpectationDto.builder()
          .currentStar(0)
          .targetStar(0)
          .isNoljang(false)
          .costWithoutDestroyPrevention(BigDecimal.ZERO)
          .costWithoutDestroyPreventionText("0")
          .expectedDestroyCountWithout(BigDecimal.ZERO)
          .costWithDestroyPrevention(BigDecimal.ZERO)
          .costWithDestroyPreventionText("0")
          .expectedDestroyCountWith(BigDecimal.ZERO)
          .build();
    }
  }

  @Getter
  @Builder
  @Jacksonized
  public static class FlameExpectationDto {
    private final BigDecimal powerfulFlameTrials;
    private final BigDecimal eternalFlameTrials;
    private final BigDecimal abyssFlameTrials;

    public static FlameExpectationDto empty() {
      return FlameExpectationDto.builder()
          .powerfulFlameTrials(BigDecimal.ZERO)
          .eternalFlameTrials(BigDecimal.ZERO)
          .abyssFlameTrials(BigDecimal.ZERO)
          .build();
    }
  }

  @Getter
  @Builder
  @Jacksonized
  public static class CostBreakdownDto {
    private final BigDecimal blackCubeCost;
    private final BigDecimal redCubeCost;
    private final BigDecimal additionalCubeCost;
    private final BigDecimal starforceCost;
    private final BigDecimal flameCost;

    public static CostBreakdownDto empty() {
      return CostBreakdownDto.builder()
          .blackCubeCost(BigDecimal.ZERO)
          .redCubeCost(BigDecimal.ZERO)
          .additionalCubeCost(BigDecimal.ZERO)
          .starforceCost(BigDecimal.ZERO)
          .flameCost(BigDecimal.ZERO)
          .build();
    }
  }
}

File: module-web/src/main/kotlin/maple/expectation/web/dto/v5/EquipmentExpectationResponseV5.kt
package maple.expectation.web.dto.v5

import java.math.BigDecimal
import java.time.Instant

/**
 * V5 CQRS 장비 기대값 응답 DTO
 */
data class EquipmentExpectationResponseV5(
    val userIgn: String,
    val calculatedAt: Instant,
    val fromCache: Boolean,
    val totalExpectedCost: BigDecimal,
    val totalCostText: String,
    val totalCostBreakdown: CostBreakdownDto,
    val maxPresetNo: Int,
    val presets: List<PresetExpectation>,
) {
    data class PresetExpectation(
        val presetNo: Int,
        val totalExpectedCost: BigDecimal,
        val totalCostText: String,
        val costBreakdown: CostBreakdownDto,
        val items: List<ItemExpectationV5>,
    )

    data class ItemExpectationV5(
        val itemName: String,
        val itemIcon: String,
        val itemPart: String,
        val itemLevel: Int,
        val expectedCost: BigDecimal,
        val expectedCostText: String,
        val costBreakdown: CostBreakdownDto,
        val enhancePath: String,
        val potentialGrade: String,
        val additionalPotentialGrade: String,
        val currentStar: Int,
        val targetStar: Int,
        val isNoljang: Boolean,
        val specialRingLevel: Int,
        val blackCubeExpectation: CubeExpectationDto,
        val additionalCubeExpectation: CubeExpectationDto,
        val starforceExpectation: StarforceExpectationDto,
        val flameExpectation: FlameExpectationDto,
    )

    data class CubeExpectationDto(
        val expectedCost: BigDecimal,
        val expectedCostText: String,
        val expectedTrials: BigDecimal,
        val currentGrade: String,
        val targetGrade: String,
        val potential: String,
    ) {
        companion object {
            @JvmStatic
            fun empty() = CubeExpectationDto(
                expectedCost = BigDecimal.ZERO,
                expectedCostText = "0",
                expectedTrials = BigDecimal.ZERO,
                currentGrade = "",
                targetGrade = "",
                potential = "",
            )
        }
    }

    data class StarforceExpectationDto(
        val currentStar: Int,
        val targetStar: Int,
        val isNoljang: Boolean,
        val costWithoutDestroyPrevention: BigDecimal,
        val costWithoutDestroyPreventionText: String,
        val expectedDestroyCountWithout: BigDecimal,
        val costWithDestroyPrevention: BigDecimal,
        val costWithDestroyPreventionText: String,
        val expectedDestroyCountWith: BigDecimal,
    ) {
        companion object {
            @JvmStatic
            fun empty() = StarforceExpectationDto(
                currentStar = 0,
                targetStar = 0,
                isNoljang = false,
                costWithoutDestroyPrevention = BigDecimal.ZERO,
                costWithoutDestroyPreventionText = "0",
                expectedDestroyCountWithout = BigDecimal.ZERO,
                costWithDestroyPrevention = BigDecimal.ZERO,
                costWithDestroyPreventionText = "0",
                expectedDestroyCountWith = BigDecimal.ZERO,
            )
        }
    }

    data class FlameExpectationDto(
        val powerfulFlameTrials: BigDecimal,
        val eternalFlameTrials: BigDecimal,
        val abyssFlameTrials: BigDecimal,
    ) {
        companion object {
            @JvmStatic
            fun empty() = FlameExpectationDto(
                powerfulFlameTrials = BigDecimal.ZERO,
                eternalFlameTrials = BigDecimal.ZERO,
                abyssFlameTrials = BigDecimal.ZERO,
            )
        }
    }

    data class CostBreakdownDto(
        val blackCubeCost: BigDecimal,
        val redCubeCost: BigDecimal,
        val additionalCubeCost: BigDecimal,
        val starforceCost: BigDecimal,
        val flameCost: BigDecimal,
    ) {
        companion object {
            @JvmStatic
            fun empty() = CostBreakdownDto(
                blackCubeCost = BigDecimal.ZERO,
                redCubeCost = BigDecimal.ZERO,
                additionalCubeCost = BigDecimal.ZERO,
                starforceCost = BigDecimal.ZERO,
                flameCost = BigDecimal.ZERO,
            )
        }
    }
}
